### PR TITLE
Initial support for YOLOv3-tiny, 3x performance on edge TPU

### PR DESCRIPTION
**Pull request by no2chem**
This PR adds support for YOLOv3-tiny, which increases the detection fps vs mobiledet from about 9fps to 27 fps on my Edge TPU. At least in my tests, accuracy seems to be pretty similar, if not slightly better.

The model used is located [here](https://github.com/guichristmann/edge-tpu-tiny-yolo).

There are some differences in the tensor output of the YOLO models, so I've refactored things out to a separate function. Hopefully that makes adding different models easier in the future. I've taken a large chunk of code from @guichristmann's repository and adapted it. I'm not exactly a numpy wizard, so there's some stuff left over relating to image scaling that can probably be removed. 

I've added some documentation on how to use YOLOv3-tiny as well. 

You'll need to map the model in via a volume mount, and use the following model configuration:

```yaml
model:
  # Required: height of the trained model
  height: 416
  # Required: width of the trained model
  width: 416
  # Required: type of model (ssd or yolo)
  model_type: 'yolo'
  # Required: path of label map
  label_path: '/labelmap.txt'
  # Optional: (but required for yolo) - anchors, comma separated
  anchors:  '10,14,  23,27,  37,58,  81,82,  135,169,  344,319'
```

I'm working on trying to quantize YOLOv4-tiny to get that working on the Edge TPU next.

### CODE REVIEW

**Comment by digiblur** [comment]
Nice!  I'll have to check this one out!

**Comment by rogerquake** [comment]
Just tried it out, and didn't detect me right away even when I was pretty close in the daylight. Eventually it got me, but it took way longer than the built-in models. Any ideas for other tweaks that are required to make this function properly?

**Comment by no2chem** [comment]
Interesting, @rogerquake 

It probably is functioning properly, I assume you were running on the Edge TPU? What kind of detection rates were you getting?

I'm checking out the mAP in an attempt to quantify the difference objectively.

**Comment by extremeshok** [comment]
Any progress with this PR ? What needs to be done to have it as an option ? ie user can select yolo or mobiledect

**Comment by jonwilliams84** [comment]
I like the idea of this...would also be useful to enable it to automatically switch to a Night Time / IR optimised model when a camera switches to b/w to get better accuracy at night?

**Comment by andrempo** [comment]
Do you think this would be better for CPU too?

**Comment by ozett** [comment]
> Do you think this would be better for CPU too?

i did not measure my yolo code on CPU, but ist running over a year and did not notice any burdon on the CPU,
abeit its only doing inference on motions sequences, not constantly.

but if i have wild guess, i would think its resource-friendly...

**Comment by gururise** [comment]
Very excited about YoloV4-tiny, but really curious to try Yolo V3

**Comment by osos** [comment]
yolov5 ?
https://blog.roboflow.com/yolov5-v6-0-is-here/

**Comment by ozett** [comment]
> Very excited about YoloV4-tiny, but really curious to try Yolo V3

do you have yolov4-tiny running with frigate? some instructions to get started? 
greatly interested to to try this in comparison...

**Comment by ozett** [comment]
YOLOv5 could somehow run as an edgeTPU model now...
anybody got into this?

https://github.com/ultralytics/yolov5/pull/3630#issuecomment-976300814

**Comment by triptec** [comment]
@ozett there's also a [export.py](https://github.com/ultralytics/yolov5/blob/master/export.py) in the repo.
I'm currently moving from pytorch+yolov5s to frigate. I'm only using cpu for detection and would really like to yolo instead because the difference in detection between ssd and yolov5 for my use case(detect cats/dogs) feels quite big (I've only done detection on the whole frame with yolo so I'm interested to see what happens if it only got the area of interest.

@blakeblackshear any thoughts on this PR or on how to be able to run other models like yolo for example?

**Comment by triptec** [comment]
@ozett actually I've managed to integrate yolov5s with smaller images. I don't have a tpu so don't know if that works but I'm getting detection fps around 5-9. And currently it seems one can't event get a hold of a usb tpu.

**Comment by ozett** [comment]
> @ozett actually I've managed to integrate yolov5s

thats great news. did you documented it anywhere? github?
do you still have the functionality of frigate post-detections?
i thought that responseformats or detections-messages by another modell type may break frigate.
but if that works, i may tray to use that too? I have another spare maschine with a TPU for that available..

sounds great to try this, and to try to bring the trained model in yolov5 for our cats , which is running fine stand-alone, now into frigate...

**Comment by triptec** [comment]
@ozett [here](https://github.com/blakeblackshear/frigate/pull/3099) you go =)

**Comment by blakeblackshear** [comment]
This branch is so old that these changes would need to be rewritten on current code.

**Comment by ozett** [comment]
@triptec 
here you use yolov5 as [a converted .tflite model](https://github.com/ultralytics/yolov5/issues/4586)?
not in its full form with the whole pytorch stuff?

(yolov3 was simple small, but yolov5 seems different, and yolov5.tflite is an "converted" export.
 just for clarification of the differences of the final model files)

**Comment by triptec** [comment]
@ozett correct, the models linked are exported versions of yolov5 that runs on tensorflow-lite without any pytorch stuff.

> (yolov3 was simple small, but yolov5 seems different, and yolov5.tflite is an "converted" export. just for clarification of the differences of the final model files)

I'm not sure what you mean. But I think yolov3 was built with tensorflow while yolov5 was built with pytorch. I think for any of them to work on tensorflow-lite/tpu they need to be converted. But I am no expert. I'm also not sure by what measure yolov3 was small and simple where yolov5 is not. I think it depends on what version of the models you select.

**Comment by svenstaro** [comment]
@blakeblackshear same as #4532 this is way out of date.
